superset-frontend/src/explore/components/ExploreViewContainer.jsx

```diff
@@ -146,11 +147,11 @@ const getWindowSize = () => ({
 function useWindowSize({ delayMs = 250 } = {}) {
   const [size, setSize] = useState(getWindowSize());
 
-  useEffect(() => {
+  useComponentDidMount(() => {
     const onWindowResize = debounce(() => setSize(getWindowSize()), delayMs);
     window.addEventListener('resize', onWindowResize);
     return () => window.removeEventListener('resize', onWindowResize);
-  }, []);
+  });
```

Comment:
I am wary of `useComponentDidMount` making it easier to ignore dependencies. Hooks are a new style of business logic organization that focuses on data flow rather than component lifecycles. Emulating the old component lifecycle focused design and silencing lint warnings is a step backwards in my view.

In this particular case, for example, rather than switching to `useComponentDidMount`, it would probably be more correct here to add `delayMs` to the dependencies array.

I do think there could be a place for `useComponentDidMount`, but only when the logic in question specifically depends on only running at component mount.

Reply:
+1 on not adding hacks just for suppressing warnings. These lint rules exist for a reason.

We should fix the dependencies rather than suppressing them.

Reply:
@suddjian I understand your point, however I did not refactor this file (if I had even done it I would have passed it to typescript).
I just kept the same logic / behavior that already existed, if "this behavior is bad" I believe it would be a matter for another PR.

Reply:
@ktmud and @suddjian about the warning.

This ``[]`` causes a warning! And the only way to eliminate it is by adding  `` // eslint-disable-next-line react-hooks/exhaustive-deps `` unfortunately it cannot be avoided.
Even in cases where it is desired to be executed only once, it is necessary to put "eslint-disable"... Of course, lint rules exist for a reason, but again, in this case the rule is not smart enough to understand that the desired behavior is exactly that.

So there are two choices, 1- put "eslint disable" in all the places where we want to execute the effect only once (currently we have something close to 7 files like this) or 2- centralize "eslint-disable" and reduce this as much as possible.

It seems to me that not adopting such a hook reinforces the use of "eslint-disable"... Which for me is a much worse practice.

Reply:
There are extremely complex class components in the application that use ComponentDidMount. Refactoring this and removing this behavior in one step will be extremely difficult and at a very high risk.

So it is absolutely not true that hooks are "a new style of business logic organization that focuses on data flow rather than component lifecycles". The main motivations are **"It's hard to reuse stateful logic between components"**, **"Complex components become hard to understand"** and **"Classes confuse both people and machines"** and they say the hooks are for: **"Hooks let you use more of React's features without classes"**.

Clearly their intention is to provide the same lifecicle/features of classes for functions, facilitate the reuse of code and reduce the size/complexity of the components.

Reply:
I think the point is that there is tech debt in these areas, and just applying `// eslint-disable-next-line react-hooks/exhaustive-deps` everywhere there is currently a warning is just further obfuscating that tech debt.

Reply:
@nytai If you need behavior similar to a "ComponentDidMount" it is a ``useEffect( effect, [] )``. But the ``[]`` will **always** generate an error with the rule and forces you to use a ``// eslint-disable-next-line react-hooks/exhaustive-deps``.
In **this case** it is not a technical debt, it is a design decision... I am extremely against "hiding" lint errors (and I hate use "eslint-disable"). But in this case, this is exactly the desired behavior, there is no escaping for it (I already tried, and a lot).

See the "FlashProvider", this is the desired behavior (apply the effect **exclusively** once).
The cases where I subsisted useEffect were the cases that it was being applied in this way.  The hook is not wrong, it is not a technical debt and it is not an "anti pattern". If I applied it to a file that already contained the same logic and it was wrong it is not a problem with the hook or with this PR, it is a problem with the PR that accepted the file with the "bad implementation"

Reply:
Right, and having the warnings thrown is a constant reminder to revisit some of that tech debt. Not every PR can be perfect the first time around. In the case of this change, https://github.com/apache/superset/pull/13220/files#diff-9c9be0048f0d0d4f1bd547faad9f8732fe5a12df6fc185190681914251131cc9R67 I think the more correct behavior would be to add `infoEnable` and `resource` to the dependency array. It seems by applying your `useComponentDidMount` to that code in this PR, you are asserting that that is the current behavior is correct.

Reply:
Let's do this way. I will revert this commit where I put this hook in these other components. I didn't really read them, I just saw that they was using it that way and switched to hook... I just made this change to avoid repeated code. As it has already led to all this discussion, it has lost its purpose.

The hook is ok, the component is ok. It does not make sense to keep blocking the merge due to other components that are not the scope of this PR. Do you agree with that?

Reply:
It's a right call to make each PR independent and not changing unrelated things unless necessary.

Reply:
@yardz Please refer to this post for why ignoring the exhaustive-deps warning is a bad idea: https://kentcdodds.com/blog/react-hooks-pitfalls#pitfall-2-not-using-or-ignoring-the-eslint-plugin

In my experience, most places where `useEffect` with an empty dependency list are used, it's either very easy to add the exhaustive deps without adding unnecessary re-renderings, or more correct to do so:

- `DynamicPlugins` : ESLint is not reporting error
- `useWindowSize`: it's more accurate to add `delayMs` because if you pass a different `delayMs`, it should add a new event handler (and remove the previous one).
- `ExploreViewContainer`: dependencies should be added as recommended, handler functions should be wrapped with `useCallback`. Because when the component get re-rendered, the old handlers would refer to closured variables in previously rendering cycle.
- `useListViewResource`: dependencies should be added because if the component that calls the hook re-renders and requests a different resource or passes in a different handler, there will be no API requests made. This makes the hook functionally incorrect. The correct way to avoid excessive requests is to cache API requests in another layer (either another hook or a new API client with cache support).
- `Welcome.tsx`: the only real dependency is `user_id`, adding it will not cause excessive re-rendering, but will also make it possible to re-render the page for different users (without unmounting the component).

Fixing the lint error is also very easy in most cases, you can simply hit "Cmd + ." in VSCode and use the recommended fix:

![image](https://user-images.githubusercontent.com/335541/108483108-2ee9f700-724f-11eb-85f3-f5ee990111f0.png)

Reply:
I think removing the unrelated changes would be best. I'm okay with having a `useComponentDidMount` hook around, as long as it is only used with code that is specifically dependent on that life cycle event, and isn't just using it to avoid lint warnings. A good example of this is the LOG_ACTIONS_MOUNT_EXPLORER event, which is concerned specifically the explorer container being mounted. I expect legit cases of `useComponentDidMount` to be very rare.

A side note on the tech debt in `ExploreViewContainer`:

A lot of the warnings here are left over from when I converted it from a class to a function component. I didn't have time to buff out all the edges there. I'm confident that better organization of state and business logic in this component would simplify the code and get rid of the warnings. For example, the `addHistory` function is called from multiple places, because in the class component we had to handle changes in that imperative way. But I suspect this functionality could instead be placed in a `useEffect` hook that sets the url based on state changes, which would be more declarative and robust.

That is what I was getting at when I said hooks are a new state management pattern that isn't based on component lifecycles. KCD discusses this in the post that @ktmud linked to as well.

Reply:
I put it just thinking about avoiding repeated code. I reversed the commit and didn't apply it anywhere else.

I agree that the hook will be little used, but I am of the opinion that if it is used 2 times, then it should be isolated in a file and reused (it is better for testing, avoiding bugs, etc.).